File: JavaScript/ExampleTests/common/ApiStatusString.js
import * as tokens from "./tokens.js";

export class ApiStatusString {
  static DNC_STRING = "?DNC?";

  constructor(code = "") {
    this.code = code;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ApiStatusString(
      v[0] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 1, "ApiStatus");
    return new ApiStatusString(
      parts[0]
    );
  }

  toString() {
    return tokens.token(this.code);
  }

  equals(other) {
    if (!(other instanceof ApiStatusString)) return false;
    return (this.code === ApiStatusString.DNC_STRING || other.code === ApiStatusString.DNC_STRING || this.code === other.code);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ApiStatusString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class ApiStatusString
    {
        public string code;

        public ApiStatusString(string code)
        {
            this.code = code;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static ApiStatusString FromText(string text)
        {
            var parts = Tokens.Require(text, 1, "ApiStatus");
            return new ApiStatusString(parts[0]);
        }

        public ApiStatusTyped ToApiStatusTyped()
        {
            return new ApiStatusTyped(
                int.Parse(this.code)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(code);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not ApiStatusString other) return false;
            return (DNCString == this.code || DNCString == other.code || object.Equals(this.code, other.code));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.code);
            return h.ToHashCode();
        }
    }
}
